clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -1764,7 +1765,15 @@ LValue CIRGenFunction::buildCastLValue(const CastExpr *E) {
     assert(0 && "NYI");
   }
   case CK_AddressSpaceConversion: {
-    assert(0 && "NYI");
+    LValue LV = buildLValue(E->getSubExpr());
+    QualType DestTy = getContext().getPointerType(E->getType());
+    mlir::Value V = getTargetHooks().performAddrSpaceCast(
+        *this, LV.getPointer(), E->getSubExpr()->getType().getAddressSpace(),
+        E->getType().getAddressSpace(), ConvertType(DestTy));
+    assert(!UnimplementedFeature::tbaa());
```

Comment:
Should watch out for the possible race condition when merging #649 . The checks may pass before the refactor but fail to compile after.